### PR TITLE
Implement Smudge classes

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/platform/w3dengine/client/w3dsmudge.cpp [hunk 1]
```diff
@@ -13,7 +13,404 @@
...
+        indices[4] = index + 4;
+        indices[5] = index + 2;
+        indices[6] = index + 2;
+        indices[7] = index + 4;
+        indices[8] = index + 1;
+        indices[9] = index + 1;
+        indices[10] = index + 4;
+        indices[11] = index;
+        index += 5;
+        indices += 12;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dsmudge.cpp]: Would be nice to put some constexpr variable(s) in file that hold the value relationship of 12 * 500 = 6000 with m_indexBuffer. I see the 500 is used in other function too.

File: src/platform/w3dengine/client/w3dsmudge.cpp [hunk 2]
```diff
@@ -13,7 +13,404 @@
...
+void W3DSmudgeManager::Release_Resources()
+{
+    Ref_Ptr_Release(m_indexBuffer);
+}
+
+void W3DSmudgeManager::Re_Acquire_Resources()
+{
+#ifdef BUILD_WITH_D3D8
+    Release_Resources();
+    SurfaceClass *buffer = DX8Wrapper::Get_DX8_Back_Buffer();
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dsmudge.cpp:56]: `Get_DX8_Back_Buffer` can return null. Needs no checking here?

File: src/platform/w3dengine/client/w3dsmudge.cpp [hunk 3]
```diff
@@ -13,7 +13,404 @@
...
+        device->DrawPrimitiveUP(D3DPT_TRIANGLESTRIP, 2, (char *)vertices, sizeof(HardwareTestVertex));
+        size = Copy_Rect(dest2, 256, 0, 0, 8, 8);
+
+        if (!size) {
+            m_hardwareTestResult = 1;
+            return false;
+        }
+
+        if (!memcmp(dest2, dest, size)) {
+            m_hardwareTestResult = 2;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dsmudge.cpp]: 2 seems to mean SUCCESS and 1 FAILURE. Can we have an enum for this?

File: src/platform/w3dengine/client/w3dsmudge.cpp [hunk 4]
```diff
@@ -13,7 +13,404 @@
...
+                                        vertices->y = c->m_position.Y;
+                                        vertices->z = c->m_position.Z;
+                                        vertices->nx = 0.0;
+                                        vertices->ny = 0.0;
+                                        vertices->nz = 0.0;
+                                        vertices->diffuse = color[k];
+                                        vertices->u1 = c->m_uv.X;
+                                        vertices->v1 = c->m_uv.Y;
+                                        vertices->u2 = 0.0;
+                                        vertices->v2 = 0.0;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dsmudge.cpp]: 0.0f

File: src/platform/w3dengine/client/w3dsmudge.cpp [hunk 5]
```diff
@@ -13,7 +13,404 @@
...
+                float f3 = f1 * 0.5f;
+                float f4 = f2 * 0.5f;
+                SmudgeSet *set = m_smudgeList.Head();
+                int count = 0;
+
+                if (set) {
+                    SortingRendererClass::Flush();
+                }
+
+                while (set) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dsmudge.cpp]: Style: `while (set != nullptr)`

File: src/platform/w3dengine/client/w3dsmudge.cpp [hunk 6]
```diff
@@ -13,7 +13,404 @@
...
+                            (smudge->m_smudgeRandom.X + 0.5) * f5 + smudge->m_smudgeStructs->m_uv.X;
+                        smudge->m_smudgeStructs[4].m_uv.Y =
+                            (smudge->m_smudgeRandom.X + 0.5) * f6 + smudge->m_smudgeStructs->m_uv.Y;
+                        count++;
+                    }
+
+                    set = set->Succ();
+                }
+
+                if (count) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dsmudge.cpp]: Style: `count != 0`

File: src/platform/w3dengine/client/w3dsmudge.cpp [hunk 7]
```diff
@@ -13,7 +13,404 @@
...
+                        }
+
+                        int count3 = 0;
+                        DynamicVBAccessClass vb(VertexBufferClass::BUFFER_TYPE_DYNAMIC_DX8, DX8_FVF_XYZNDUV2, 5 * count);
+
+                        {
+                            DynamicVBAccessClass::WriteLockClass lock(&vb);
+                            VertexFormatXYZNDUV2 *vertices = lock.Get_Formatted_Vertex_Array();
+
+                            while (set) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dsmudge.cpp]: Style: `while (set != nullptr)`

File: src/platform/w3dengine/client/w3dsmudge.cpp [hunk 8]
```diff
@@ -13,7 +13,404 @@
...
+                                    color2 = ((unsigned int)(s->m_smudgeAlpha * 255.0f) << 24) | 0xFFEEDD;
+
+                                    for (int k = 0; k < 5; k++) {
+                                        vertices->x = c->m_position.X;
+                                        vertices->y = c->m_position.Y;
+                                        vertices->z = c->m_position.Z;
+                                        vertices->nx = 0.0;
+                                        vertices->ny = 0.0;
+                                        vertices->nz = 0.0;
+                                        vertices->diffuse = color[k];
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dsmudge.cpp:382]: `color[4]` has 4 elements, but k can access 5. This looks wrong?

File: src/platform/w3dengine/client/w3dsmudge.cpp [hunk 9]
```diff
@@ -13,7 +13,404 @@
...
+                }
+
+                while (set) {
+                    DLListClass<Smudge> &list = set->Get_Smudge_List();
+
+                    for (Smudge *smudge = list.Head(); smudge != nullptr; smudge = smudge->Succ()) {
+                        Matrix3D::Transform_Vector(m2, smudge->m_smudgePos, &v1);
+                        smudge->m_smudgeStructs[4].m_position = v1;
+
+                        for (int i = 0; i < 4; i++) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dsmudge.cpp]: `i < ARRAY_SIZE(v3)` maybe

File: src/platform/w3dengine/client/w3dsmudge.cpp [hunk 10]
```diff
@@ -13,7 +13,404 @@
...
+#ifdef BUILD_WITH_D3D8
+    if (Test_Hardware_Support()) {
+        CameraClass &camera = rinfo.m_camera;
+        Vector3 v1;
+        Vector4 v2;
+
+        Vector3 v3[4] = {
+            Vector3(-0.5f, 0.5f, 0.0f), Vector3(-0.5f, -0.5f, 0.0f), Vector3(0.5f, -0.5f, 0.0f), Vector3(0.5f, 0.5f, 0.0f)
+        };
+
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dsmudge.cpp:259]: `static_assert(ARRAY_SIZE(v3) <= ARRAY_SIZE(Smudge::m_smudgeStructs)`

File: src/platform/w3dengine/client/w3dsmudge.cpp [hunk 11]
```diff
@@ -13,7 +13,404 @@
...
+                                smudge->m_smudgeRandom.Y = 0.0f;
+                            }
+                        }
+
+                        float f5 = smudge->m_smudgeStructs[3].m_uv.X - smudge->m_smudgeStructs->m_uv.X;
+                        float f6 = smudge->m_smudgeStructs[1].m_uv.Y - smudge->m_smudgeStructs->m_uv.Y;
+                        smudge->m_smudgeStructs[4].m_uv.X =
+                            (smudge->m_smudgeRandom.X + 0.5) * f5 + smudge->m_smudgeStructs->m_uv.X;
+                        smudge->m_smudgeStructs[4].m_uv.Y =
+                            (smudge->m_smudgeRandom.X + 0.5) * f6 + smudge->m_smudgeStructs->m_uv.Y;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dsmudge.cpp]: 0.5f

File: src/platform/w3dengine/client/w3dsmudge.cpp [hunk 12]
```diff
@@ -13,7 +13,404 @@
...
+                int width = desc.Width;
+                int height = desc.Height;
+                float f1 = (float)g_theTacticalView->Get_Width() / (float)width;
+                float f2 = (float)g_theTacticalView->Get_Height() / (float)height;
+                float f3 = f1 * 0.5f;
+                float f4 = f2 * 0.5f;
+                SmudgeSet *set = m_smudgeList.Head();
+                int count = 0;
+
+                if (set) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dsmudge.cpp]: Style: `while (set != nullptr)`

File: src/platform/w3dengine/client/w3dsmudge.cpp [hunk 13]
```diff
@@ -13,7 +13,404 @@
...
+                            while (set) {
+                                for (Smudge *s = smudge; s; s = s->Succ()) {
+                                    SmudgeCoords *c = s->m_smudgeStructs;
+
+                                    if (count3 >= count) {
+                                        smudge = s;
+                                        goto done;
+                                    }
+
+                                    color2 = ((unsigned int)(s->m_smudgeAlpha * 255.0f) << 24) | 0xFFEEDD;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dsmudge.cpp]: `color2` has value assigned here, but it appears to be used no where?
I assume color2 is part of `color[5]` array.

Comment (jonwil) [reply src/platform/w3dengine/client/w3dsmudge.cpp]: fixed.

File: src/platform/w3dengine/client/w3dsmudge.cpp (hunk 12 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dsmudge.cpp]: fixed.

File: src/platform/w3dengine/client/w3dsmudge.cpp (hunk 11 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dsmudge.cpp]: fixed.

File: src/platform/w3dengine/client/w3dsmudge.cpp (hunk 9 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dsmudge.cpp]: fixed.

File: src/platform/w3dengine/client/w3dsmudge.cpp (hunk 6 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dsmudge.cpp]: fixed.

File: src/platform/w3dengine/client/w3dsmudge.cpp (hunk 7 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dsmudge.cpp]: fixed.

File: src/platform/w3dengine/client/w3dsmudge.cpp (hunk 8 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dsmudge.cpp:382]: fixed.

File: src/platform/w3dengine/client/w3dsmudge.cpp (hunk 5 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dsmudge.cpp]: fixed.

File: src/platform/w3dengine/client/w3dsmudge.cpp (hunk 4 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dsmudge.cpp]: fixed.

File: src/platform/w3dengine/client/w3dsmudge.cpp (hunk 2 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dsmudge.cpp:56]: Original code had no checking.

File: src/platform/w3dengine/client/w3dsmudge.cpp (hunk 10 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dsmudge.cpp:259]: fixed.

File: src/platform/w3dengine/client/w3dsmudge.cpp (hunk 1 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dsmudge.cpp]: Added some appropriate enums for this based on how other similar classes work.

File: src/platform/w3dengine/client/w3dsmudge.cpp (hunk 3 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dsmudge.cpp]: Added an enum for this.

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**